src/main/java/com/github/black0nion/blackonionbot/rest/impl/post/Login.java

```diff
@@ -1,29 +1,66 @@
 package com.github.black0nion.blackonionbot.rest.impl.post;
 
-import com.github.black0nion.blackonionbot.oauth.DiscordUser;
-import com.github.black0nion.blackonionbot.oauth.OAuthHandler;
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.algorithms.Algorithm;
+import com.github.black0nion.blackonionbot.database.helpers.api.SQLHelperFactory;
+import com.github.black0nion.blackonionbot.oauth.OAuthUser;
 import com.github.black0nion.blackonionbot.rest.api.IPostRoute;
 import com.github.black0nion.blackonionbot.rest.sessions.RestSession;
-import com.github.black0nion.blackonionbot.utils.Utils;
 import io.javalin.http.BadRequestResponse;
 import io.javalin.http.Context;
+import io.mokulu.discord.oauth.DiscordAPI;
+import io.mokulu.discord.oauth.DiscordOAuth;
+import io.mokulu.discord.oauth.model.TokensResponse;
 import org.jetbrains.annotations.Nullable;
 import org.json.JSONObject;
 
 import javax.annotation.Nonnull;
-import java.util.InputMismatchException;
+import java.time.Instant;
+
+import static com.github.black0nion.blackonionbot.utils.AuthUtils.JWT_VALID_FOR;
 
 public class Login implements IPostRoute {
 
-	private final OAuthHandler handler;
+	private final Algorithm algorithm;
+	private final DiscordOAuth discordOAuth;
+	private final SQLHelperFactory sqlHelperFactory;
 
-	public Login(OAuthHandler handler) {
-		this.handler = handler;
+	public Login(Algorithm algorithm, DiscordOAuth discordOAuth, SQLHelperFactory sqlHelperFactory) {
+		this.algorithm = algorithm;
+		this.discordOAuth = discordOAuth;
+		this.sqlHelperFactory = sqlHelperFactory;
 	}
 
 	@Override
-	public Object handle(Context ctx, JSONObject body, @Nullable RestSession session, DiscordUser user) throws Exception {
-		return Utils.replaceException(() -> handler.loginWithDiscord(ctx.header("code")), InputMismatchException.class, BadRequestResponse.class);
+	public Object handle(Context ctx, JSONObject body, @Nullable RestSession session, OAuthUser user) throws Exception {
+		String code = ctx.queryParam("code");
+		if (code == null) {
+			throw new BadRequestResponse("Missing code parameter");
+		}
+
+		long requestTime = System.currentTimeMillis();
+		TokensResponse tokens = discordOAuth.getTokens(code);
+		String accessToken = tokens.getAccessToken();
+		String refreshToken = tokens.getRefreshToken();
+		long expiresAt = requestTime / 1000 + tokens.getExpiresIn();
+
+		user = new OAuthUser(accessToken, refreshToken, expiresAt, new DiscordAPI(accessToken));
+
+		sqlHelperFactory.run("INSERT INTO sessions (user_id, access_token, refresh_token, expires_at) VALUES (?, ?, ?, ?) ON CONFLICT (user_id) DO UPDATE SET access_token = ?, refresh_token = ?, expires_at = ?",
+			user.getId(),
+			accessToken,
+			refreshToken,
+			expiresAt,
+			accessToken,
+			refreshToken,
+			expiresAt
+		);
+
+		return JWT.create()
+			.withExpiresAt(Instant.now().plusSeconds(JWT_VALID_FOR))
```

Comment:
The start of the validity period has to be computed from `requestTime` rather than `Instant.now()` (the SQL execution could take a while)

Reply:
Actually, nevermind, that's *our* token with *our* expiration time and not connected to discord directly